### PR TITLE
fix(nayduck): State sync skip state dump

### PR DESCRIPTION
**Pull request by VanBarbascu**
This addresses `sanity/state_sync_fail.py` and `sanity/gc_after_sync.py`.

`sanity/state_sync_fail.py`- the node was trying to dump after the shard layout change.
`sanity/gc_after_sync.py` did not have state sync configured and the epoch was to short for the node to sync to the right epoch.

### CODE REVIEW

**Review by Longarithm — APPROVED** [review]
Thank you! Approving to unblock, but please fix newly failing tests in https://nayduck.near.org/#/run/3433 (sanity/state_sync.py manytx 30...). Also maybe rename `state_sync_fail` to `state_sync_resharding_fail` everywhere to make the reason of failing more clear for nayduck readers.

File: core/primitives/src/state_sync.rs [hunk 1]
```diff
@@ -261,6 +261,7 @@ pub fn get_num_state_parts(memory_usage: u64) -> u64 {
 pub enum StateSyncDumpProgress {
     /// Represents two cases:
     /// * An epoch dump is complete
+    /// * An epoch dump is skipped in the epoch where shard layout changes
     /// * The node is running its first epoch and there is nothing to dump.
     AllDumped {
```
Comment (wacban) [inline core/primitives/src/state_sync.rs:266]: This has two meanings now and is confusing. I would suggest splitting it into two separate enum values. Alternatively but not as clean rename it to AllDone.

File: nearcore/src/state_sync.rs [hunk 2]
```diff
@@ -209,19 +209,31 @@ fn get_current_state(
     account_id: &Option<AccountId>,
     epoch_manager: Arc<dyn EpochManagerAdapter>,
 ) -> Result<Option<(EpochId, EpochHeight, CryptoHash)>, Error> {
```
Comment (wacban) [inline nearcore/src/state_sync.rs:211]: nit: I can't tell from the return type what's what. Can you change it to a dedicated enum with meaningful names for the available options?

File: nearcore/src/state_sync.rs [hunk 3]
```diff
@@ -594,7 +606,7 @@ fn get_latest_epoch(
     shard_id: &ShardId,
     chain: &Chain,
     epoch_manager: Arc<dyn EpochManagerAdapter>,
-) -> Result<(EpochId, EpochHeight, CryptoHash), Error> {
+) -> Result<(EpochId, EpochId, EpochHeight, CryptoHash), Error> {
```
Comment (wacban) [inline nearcore/src/state_sync.rs:609]: nit: I would consider a dedicated struct for this, especially given you habe two values with the same type.

**Review by wacban — APPROVED** [review]
LGTM thanks for looking into it!

File: nearcore/src/state_sync.rs [hunk 4]
```diff
@@ -202,26 +202,50 @@ fn select_random_part_id_with_index(parts_to_be_dumped: &Vec<u64>) -> (u64, usiz
...
+
 fn get_current_state(
     chain: &Chain,
     shard_id: &ShardId,
     shard_tracker: &ShardTracker,
     account_id: &Option<AccountId>,
     epoch_manager: Arc<dyn EpochManagerAdapter>,
-) -> Result<Option<(EpochId, EpochHeight, CryptoHash)>, Error> {
-    let was_last_epoch_dumped = match chain.chain_store().get_state_sync_dump_progress(*shard_id) {
+) -> Result<Option<StateDumpShardInfo>, Error> {
```
Comment (wacban) [inline nearcore/src/state_sync.rs:217]: I was actually suggesting to remove the option and instead have a new enum for this. But ff you can just explain what does None return value mean in a comment that's good enough for me too :)

File: nearcore/src/state_sync.rs [hunk 5]
```diff
@@ -202,26 +202,50 @@ fn select_random_part_id_with_index(parts_to_be_dumped: &Vec<u64>) -> (u64, usiz
...
+        Ok(StateSyncDumpProgress::Skipped { epoch_id, .. }) => Some(epoch_id),
         _ => None,
     };
 
-    match get_latest_epoch(shard_id, &chain, epoch_manager) {
+    match get_latest_epoch(shard_id, &chain, epoch_manager.clone()) {
         Err(err) => {
             tracing::debug!(target: "state_sync_dump", shard_id, ?err, "check_latest_epoch failed. Will retry.");
             Err(err)
         }
```
Comment (wacban) [inline nearcore/src/state_sync.rs:228]: You can flatten this by handling the error upfront like so:
```
    let latest_epoch_info = get_latest_epoch(shard_id, &chain, epoch_manager);
    let latest_epoch_info = latest_epoch_info.map_err(|err| {
        tracing::error!(target: "state_sync_dump", shard_id, ?err, "Failed to get the latest epoch");
        err
    })?;
    let LatestEpochInfo{prev_epoch_id, new_epoch_id, new_epoch_height, new_sync_hash} = latest_epoch_info;
```

File: nearcore/src/state_sync.rs (hunk 4 above)
Comment (VanBarbascu) [reply nearcore/src/state_sync.rs]: Changed the result to `Dump{...}` or `Wait`. This should explain it.

**Comment by VanBarbascu** [comment]
Currently this fixes:
```
< "pytest sanity/gc_after_sync.py --features nightly"
< "pytest sanity/gc_after_sync.py"
< "pytest sanity/gc_after_sync1.py --features nightly"
< "pytest sanity/gc_after_sync1.py"
< "pytest sanity/state_sync.py onetx 30 --features nightly"
< "pytest sanity/state_sync.py onetx 30"
< "pytest sanity/state_sync_fail.py"
```
But breaks `> "pytest sanity/state_sync.py manytx 30"` for some reason. From the runs on the main branch, it seems like it is flaky.

I will land these changes and keep an eye on it.

**Pull request merged**